Metrics table generation from yaml

jamesmoessis opened the pull request:
This PR adds functionality such that metrics can be parsed from yaml and relevant information generated in markdown. 

~~This can regenerate the structure of the [http metric semantic convention](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md) that we have today, including the "Attribute alternatives" which can be controlled via the `constraints` in the yaml.~~

There are two types of tables that must be generated from `MetricSemanticConvention`: 
- attributes (same as before)
- metrics (id, instrument, units)

`<!-- semconv metric.http.client -->` - creates an attribute table as normal.

`<!-- semconv metric.http.client(metric_table) -->` - creates a metric table of all `metrics` that are under `metric.http.client` in the yaml file. If you have other ideas about how the yaml file should be structured, let me know (see `http_metrics.yaml`).

The `metric` prefix ID is used to prevent clash with the tracing semconvs of the same ID. 

It's quite few lines of actual code, most of this PR is testing. 

cc @ocelotl @jsuereth

Oberon00 commented on semantic-conventions/src/tests/semconv/model/test_correct_parse.py:
```diff
@@ -669,6 +735,12 @@ def semantic_convention_check(self, s, expected):
         self.assertEqual(expected["attributes"], [a.fqn for a in s.attributes])
 
+    def metric_check(self, metrics: Tuple[MetricSemanticConvention.Metric], expected):
+        for i, metric in enumerate(metrics):
```
You can probably use `zip(metrics, expected)` instead of `enumerate` here.

jamesmoessis replied:
Yeah that's cleaner, thanks for the suggested. I went ahead and implemented it.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -66,7 +68,7 @@ class MarkdownRenderer:
     p_end = re.compile("<!--\\s*endsemconv\\s*-->")
     default_break_conditional_labels = 50
-    valid_parameters = ["tag", "full", "remove_constraints"]
+    valid_parameters = ["tag", "full", "remove_constraints", "metric_table"]
```
Is this needed as a parameter? Shouldn't the markdown generator be able to determine this automatically based on the semantic convention group kind?

Oberon00 reviewed (commented):
All features require documentation in https://github.com/open-telemetry/build-tools/blob/main/semantic-conventions/syntax.md. This would also be quite useful to review the whole feature.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -414,15 +434,24 @@ def _render_group(self, semconv, parameters, output):
             output.write("The event name MUST be `{}`.\n\n".format(semconv.name))
 
+        if (
+            isinstance(semconv, MetricSemanticConvention)
```
Probably an error should be thrown if `is_metric_table` is set on another type of semantic convention.

jamesmoessis replied:
Good point, I've implemented that.

Oberon00 commented on semantic-conventions/src/tests/data/markdown/metrics_tables/http.yaml:
```diff
@@ -0,0 +1,172 @@
+          # Default value: `true`. If false, it helps the code gen tool to
+          # encode checks that only accept the listed values.
+          allow_custom_values: true
+          members:
```
I know that most tests have this copy&pasta, but for new tests, maybe try to keep the YAML files minimal for what you want to test.

jamesmoessis replied:
Ok, I've kept it as minimal as I can for now. I've still had to duplicate `http.yaml` once, because the main one is outdated, and I wanted to use the latest one from the spec. If I update it to the latest ones lots of other tests break.

I had to add a few lines of code to specify extra files (rather than directories) to achieve this too. Other tests can follow suit in later PRs if we want to trim the copy+pasta.

Oberon00 replied:
Fine with me, though long-term I'd love to have only foo/bar-style actually minimal YAML in these tests (we could have dedicated integration tests with spec YAML, maybe even pulling from spec repo), but that is a separate issue.

Oberon00 commented on semantic-conventions/src/tests/data/markdown/metrics_tables/http_metrics.yaml:
```diff
@@ -0,0 +1,65 @@
+      - ref: net.peer.port
+      - ref: net.peer.ip
+    metrics:
+      - id: metric.http.client.duration
```
According to the usual rules of id + prefix, the fullly qualified name of this would be "metric.http.metric.http.client.duration", with a duplicated prefix.

jamesmoessis replied:
Interesting. While I'm all for following the same patterns and rules, "metric.http.metric.http.client.duration" seems to be confusingly and needlessly long. Is there another way?

Oberon00 replied:
Of course that name would be unacceptable. Thus, you'd change the line here to just `client.duration` and the prefix is in the `prefix`. That's how it's meant to be used.

Oberon00 replied:
Compare the attributes, there you have `id: status_code` instead of `id: http.status_code`, as the prefix is prepended by the generator.

jamesmoessis replied:
I see, thanks for clearing that up. I agree we should follow the same pattern for metric names as we do for attributes. 

I've changed the `prefix` to be `metric.http` and then (for example) the metric id to `client.duration`, so the FQN is `metric.http.client.duration` as desired. Same pattern follows for other metrics.

jamesmoessis commented:
Thanks @Oberon00, I've pushed up syntax changes. Once we agree on a syntax I can add it to the JSON config too.

jamesmoessis commented:
Hey @Oberon00 just bumping on this since it's been a while. 

I feel like the main thing needing to be discussed is the structure of the metrics in the yaml and how the `extend`ing works. What else can I do to get this PR moving?

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -174,6 +176,24 @@ def to_markdown_attr(
+        This method renders metrics as markdown table entry
+        """
+        output.write(
+            "| Name     | Instrument       | Unit ([UCUM](README.md#instrument-units)) | Description    |\n"
```
This relative link URL won't work if there is no README.md in the same folder. This is currently a limitation of this tool (#20). I recommend rather not using a link, or maybe using a absolute link (having the drawback that you need to choose a branch).

jamesmoessis replied:
I was trying to regenerate what's already in the semantic convention markdown currently. 

If we need to remove the link in order to get it to work properly with the tool, I'm fine with that. So I've gone ahead and removed the link.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -414,15 +434,28 @@ def _render_group(self, semconv, parameters, output):
         for attr in sorted(
             semconv.attributes, key=lambda a: "" if a.ref is None else a.ref
         ):
+            if self.render_ctx.is_metric_table:
```
That's a bit of a hack...
Also seeing `remove_constraints` used always, I think it would make sense to separate metric_tables more from all the others. All the others are attribute tables, which the metric_table is not.

I'm thinking about something like `if "metric_table" in parameters: render_metric_group(...); else: render_attribute_group(...)`

jamesmoessis replied:
Yeah, agreed that's a better way. I've done what you've suggested, there's an if/else for rendering metrics vs attributes tables.

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -119,6 +119,18 @@ events ::= id {id} # MUST point to an existing event group
 
 name ::= string
 
+instrument ::=  "Counter" 
```
Please also document the semantics of the new syntax (below sections).

jamesmoessis replied:
Done 👍

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -53,6 +53,19 @@ def parse(span_kind_value):
 
 
+class InstrumentKind(Enum):
+    Empty = 1
```
Enum members should use ALL_CAPS

jamesmoessis replied:
👍  true, I've fixed it up. names of the instruments in the spec are CamelCase, so I had it like conveniently so I could map to `self.name`, but it wasn't too much trouble to write a mapping function to do that. 

Additionally, I've removed the asynchronous instruments as per https://github.com/open-telemetry/opentelemetry-specification/pull/2458

Oberon00 commented:
@jamesmoessis I went over the PR again, focusing on implementation & "formalities". I think for the actual high-level data model/YAML-schema, someone responsible for metrics needs to approve.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -238,7 +251,35 @@ def __init__(self, group):
+            self.brief: str = metric.get("brief")
+
+            if None in [metric.get("instrument"), self.id, self.units, self.brief]:
+                raise ValueError
```
Missing exception message.
Consider using `raise ValidationError.from_yaml_pos` like in other locations.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -238,7 +251,35 @@ def __init__(self, group):
+        self.metrics = ()
+        if group.get("metrics"):
+            try:
+                self.metrics: Tuple[MetricSemanticConvention.Metric] = tuple(
```
```suggestion
                self.metrics: Tuple[MetricSemanticConvention.Metric, ...] = tuple(
```

Otherwise the type denotes a single-element-tuple.

jamesmoessis replied:
Good catch.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -238,7 +251,35 @@ def __init__(self, group):
+                    )
+                )
+            except ValueError as e:
+                raise ValidationError.from_yaml_pos(
```
I think it would be clearer if this error was directly thrown from `Metric.__init__`.

jamesmoessis replied:
Agreed, I did it here because `Metric` doesn't have access to `_position`. Anyway, I have just passed `_position` into the constructor, so I can raise from there now.

jamesmoessis commented on semantic-conventions/syntax.md:
```diff
@@ -119,6 +119,18 @@ events ::= id {id} # MUST point to an existing event group
 
 name ::= string
 
+instrument ::=  "Counter" 
```
I was following the convention from the spec which uses CamelCase, but if it makes sense to use lowercase here then that also makes sense. I've gone ahead and done that.

jamesmoessis replied:
As discussed elsewhere, I've made this CamelCase again.

jamesmoessis commented on semantic-conventions/src/tests/semconv/templating/test_markdown.py:
```diff
@@ -133,6 +139,7 @@ def check(
         *,
         expected_name="expected.md",
         extra_yaml_dirs: Sequence[str] = (),
+        extra_yaml_files: Sequence[str] = (),
```
I wanted to use a newer `http.yaml` defined in the `metrics_tables` directory. The reason being the `yaml/http.yaml` had older data, and updating it to the newer spec yaml caused other tests to fail. I couldn't have them both parsed in the same test because I would get an error like `Semantic convention 'http' is already defined`. So it wasn't easy to parse the whole `yaml/` directory.

My solution was to be able to specify specific extra yaml files to use for a test, and this solved my problem. 

The root problem is that these test files become out of date quick, we should at some point look at pulling them directly from the spec and so it can be kept up to date more quickly. That way we don't need multiple versions of markdown and yaml laying around, and all the tests run off the same consistent data.

Oberon00 replied:
> The root problem is that these test files become out of date quick, we should at some point look at pulling them directly from the spec and so it can be kept up to date more quickly.

IMHO, no, this is not what we should do, at least not for these tests. These tests are not designed as integration tests, they are targeted feature tests and it makes no sense to drag in a full spec yaml. These tests should instead use a minimal yaml that just uses group foo with attribute bar instead of actual data. See #63.

Having tests that automatically pull in spec code would *additionally* make sense, but then we don't have e.g. a testMultipleEnum or testRef but instead a testGeneratorIsSpecCompatible (and you'd need to think about how to design the build/change process so that you can also change the output in "incompatible" ways, e.g. temporarily reference a spec PR branch that updates the spec so the semconv table check is in sync again.

----

So consider using something more minimal for this test instead of using a group named "http", maybe use a group "metric_test" or "foo".

jamesmoessis commented on semantic-conventions/dev-requirements.txt:
```diff
@@ -1,4 +1,4 @@
-black==21.8b0
+black==22.3.0
```
A transitive dependency of `black` was failing the build, updating black to latest seems to have resolved it.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -53,6 +53,32 @@ def parse(span_kind_value):
+    @staticmethod
+    def kind_map():
+        return {
+            "Counter": InstrumentKind.COUNTER,
```
I think I'd prefer to also use lowercase in the YAML, which would be consistent with everything else. Thoughts?

jamesmoessis replied:
I wanted to recreate what was already in the spec, and it seems pretty consistently using CamelCase. For example, [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#http-server). So, I've stuck with the CamelCase for now if that's ok. 

I think if we wanted to change it to lower case, that would most likely be a discussion in the spec rather than here. Any thoughts?

Oberon00 replied:
Sorry, I lost track of this PR.
The YAML property names & enums often don't follow the spec names exactly, they are all lowercase (e.g. `type: event` ). Of course, the markdown output should still be UpperCamelCase.

jamesmoessis replied:
Ok sounds good, in my recent changes I've made it so it's lowercase in the yaml but CamelCase in the markdown.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -53,6 +53,32 @@ def parse(span_kind_value):
+
+    def __str__(self):
+        # reverse lookup kind_map()
+        return next(filter(lambda i: i[1] is self, InstrumentKind.kind_map().items()))[
```
I would not override `str` for enums. What is it used for?

jamesmoessis replied:
I've gone ahead and removed this entire class. I had to do this to correlate the ENUM_NAME with the YamlName. I decided Enum is not a good fit and decided InstrumentKind doesn't need it's own type. 

It can just be a string and we check if it's allowed instrument strings.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -53,6 +53,32 @@ def parse(span_kind_value):
+
+    @staticmethod
+    def kind_map():
+        return {
```
This should be a constant instead of being re-created at every call.

jamesmoessis replied:
removed this class and method

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -416,31 +443,36 @@ def _render_group(self, semconv, parameters, output):
+        if self.render_ctx.is_metric_table:
+            self.to_markdown_metric_table(semconv, output)
+        else:
+            attr: SemanticAttribute
```
Extract the whole `else` branch to a function named `to_markdown_attribute_table`? Would probably also make the diff look nicer because the indentation would not change then.

jamesmoessis replied:
I thought about this initially but there was a soup of variables used inside and out of the branch. Anyway, I had a closer look and went ahead and did it. Things just needed a little re-arranging.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -416,31 +443,36 @@ def _render_group(self, semconv, parameters, output):
         self.render_ctx.is_full = "full" in parameters
+        self.render_ctx.is_metric_table = "metric_table" in parameters
 
         if isinstance(semconv, EventSemanticConvention):
```
This logically also belongs to the "not metrics" branch.

jamesmoessis replied:
Done 👍

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -166,6 +176,18 @@ The following is only valid if `type` is `event`:
+    semantic convention prefix like so: `{parent.prefix}.{metric.id}`. 
+  - `brief`, a brief description of the metric.
+  - `instrument`, the instrument that *should* be used to record the metric.
+  - `units`, the units in which the metric is measured.
```
Link to UCUM?

jamesmoessis replied:
Done 👍

Oberon00 reviewed (approved):
Looks OK to me now from a code perspective, I'm too out of the metrics spec to judge the semantics though. Please gather reviews from metrics-involved people on that aspect.

reyang commented on semantic-conventions/src/tests/data/markdown/metrics_tables/expected.md:
```diff
@@ -0,0 +1,76 @@
+<!-- semconv metric.http.server(metric_table,remove_constraints) -->
+| Name     | Instrument    | Unit (UCUM) | Description    |
+| -------- | ------------- | ----------- | -------------- |
+| `metric.http.server.duration` | Histogram | `ms` | Measures the duration of the inbound HTTP request. |
+| `metric.http.server.active_requests` | UpDownCounter | `{requests}` | Measures the number of concurrent HTTP requests that are currently in-flight. |
```
Could these be renamed by removing the `metric.` prefix?

jamesmoessis replied:
Yes, I agree, they could be. The `metric` prefix exists purely to stop clashes with the span semconv in code. So, we can strip it when rendering it to markdown.

jamesmoessis commented:
Thanks for the review @jmacd @reyang - I will address the comments and rebase/update this PR as soon as I get time

jamesmoessis commented on semantic-conventions/syntax.md:
```diff
@@ -166,6 +176,19 @@ The following is only valid if `type` is `event`:
+  - `id`, the ID of the metric. The fully qualified name of the metric includes its parent 
+    semantic convention prefix like so: `{parent.prefix}.{metric.id}`. 
+  - `brief`, a brief description of the metric.
+  - `instrument`, the instrument that *should* be used to record the metric.
```
Done

jamesmoessis commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -186,6 +189,62 @@ def to_markdown_attr(
+        for metric in semconv.metrics:
+            output.write(
+                "| `{}` | {} | `{}` | {} |\n".format(
+                    metric.fqn.removeprefix("metric."),
```
Yeah, you're right. That makes sense. I've pushed up the change 👍

jamesmoessis commented:
@jsuereth @arminru perhaps I can get some discussion on how the yaml and the markdown should be structured as I think that's the most important thing here. 

The hardest part I found is striking a balance between the nuances of metric attributes, as the same metric can have different attribute sets depending on the context. It makes it slightly hard to make a table. Not all `http.server.*` metrics would have the same attributes, which makes it hard to group them in a table. 

I am considering altering the approach I have here and taking some inspiration from scrapers in the collector: ([example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/72b0032d1b7be4179bc03cedd68800321386a3b4/receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/documentation.md)). The current [HTTP metrics semconv](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md) avoids listing metric attributes next to the metrics, which leads to somewhat ambiguous descriptions. Lots of metrics have attribute alternatives or optional fields. 

This means that the attributes would need to be listed separately to the metrics in the yaml, and then attached to each individual metric where necessary. This is somewhat different to the tracing approach, but I think it is the better way to do it for metrics. 

I can implement this when I get the time, but it would help this PR move along if there is some discussion and eventual agreement around how to structure the yaml/markdown for metrics.

joaopgrassi reviewed (commented):
> I am considering altering the approach I have here and taking some inspiration from scrapers in the collector: ([example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/72b0032d1b7be4179bc03cedd68800321386a3b4/receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/documentation.md)).

If I got it correctly, you mean adding a new column in the metrics table, and lay out all the attributes for each? Then have a full table, with all attributes together? I think this is a great idea and in my view would making reading much easier. 

Maybe another "nice to have" is adding anchor links to the attributes in the table so you can click around. Maybe that makes navigating between metrics table <> attributes table easier since I can then use the browser history to navigate back and forth. 

Something like:

```md
### HTTP Client Metrics

| Name     | Instrument    | Unit (UCUM) | Description    | Attributes |
| -------- | ------------- | ----------- | -------------- | -----------|
| `http.client.duration` | Histogram | `ms` | ... | [http.host](#http.host)

## Common Attributes

| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| <span id="http.host">`http.host`</span> | string | ... |... | ... |

Not sure if this is possible
```

joaopgrassi commented:
Another thing I remembered: We probably need to think if the model here will fit for code generation. I assume at some point we will want to generate code files as we do for trace semantic conventions. Then things like "how attributes are grouped" and "how they are linked to each individual metrics" makes even more sense as that might impact in how we generate code from it.

jamesmoessis commented on semantic-conventions/src/tests/data/markdown/metrics_tables/expected.md:
```diff
@@ -0,0 +1,76 @@
+### HTTP Client Metrics
+
+<!-- semconv metric.http.client(metric_table,remove_constraints) -->
+| Name     | Instrument    | Unit (UCUM) | Description    |
```
Ah yes, I think I deliberately didn't add that, to save some space. I'm not sure that two unit columns are necessarily essential, but I don't have a strong preference either.

joaopgrassi replied:
Leaving it out is fine by me, I guess most times it should be self-explanatory.

joaopgrassi commented:
>How to list attributes that are used in multiple places? For example the net.* attributes are used in http, db, anything that uses a network connection essentially.

Yes, this is a real problem with the current structure. I saw you authored this PR a while ago https://github.com/open-telemetry/opentelemetry-specification/pull/2019, which was a prerequisite for https://github.com/open-telemetry/opentelemetry-specification/pull/1977. 

Not sure there's a way to share common attributes without refactoring things like the above proposed. 

> What about the "attribute alternatives" i.e. metrics that have different sets of attributes that apply to them depending on what's available to the instrumentation

Hum.. I guess such situation are covered now with the new "Conditionally required" level? Similar to what's in here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#attributes

jamesmoessis commented:
> Hum.. I guess such situation are covered now with the new "Conditionally required" level? Similar to what's in here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#attributes

I think the problem with that document is that the attributes table tries to specify where the attribute is to be used, and if it's required, recommended, etc. This is a problem because an attribute has the potential to be used differently for each metric. For one attribute, maybe it's required for one metric, but only recommended for another. 

In order for a common structure to specify metrics and attributes scaling to all use cases, the metric table itself should specify the attributes and the requirement level, in my opinion. This can't be done for traces because the cardinality is too high, for metrics this is less of a concern. Additionally this makes the specification clearer, as each metric has a specific set of attributes listed next to it, rather than the reader having to reverse-search from the attributes table.

I believe something like the example below would work for a metric table. 

| Name                          | Instrument Type ([*](README.md#instrument-types)) | Unit ([UCUM](README.md#instrument-units)) | Description                            | Attributes             |
|-------------------------------|---------------------------------------------------|-------------------------------------------|----------------------------------------|------------------------|
| `http.server.duration`        | Histogram                                         | `ms`                                      | Measures the duration inbound HTTP requests    |<ul><li>`http.method` (Required)</li> <li>`http.host` (See attribute alternatives)</li><li>`http.scheme` (See attribute alternatives)</li> <li>`http.scheme` (See attribute alternatives)</li> <li>`http.status_code` (Conditionally required)</li> <li>`http.flavor` (Recommended)</li> <li>`http.server_name` (See attribute alternatives)</li> <li>`net.host.name` (See attribute alternatives)</li> <li>`net.host.port` (See attribute alternatives)</li> </ul> |

Then the attributes would be specified separately and anchor-linked to from here, similar to the example you posted above but with requirement level.

jamesmoessis commented:
@lmolkova, there is the question of how requirement level should be done for metrics and I was wondering if you had any thoughts. As my above comment says, I don't think it makes sense for the requirement level to be associated with the attribute definition in the case of metrics. Instead it should be associated with the metric. 

This will need a decent amount of code change to achieve, so I want to make sure there is some consensus before I go ahead and make changes.

joaopgrassi commented:
@jamesmoessis 

>In order for a common structure to specify metrics and attributes scaling to all use cases, the metric table itself should specify the attributes and the requirement level, in my opinion.

I agree and this is probably the way we can eliminate any confusion for people that are going to consume it. 

>  I don't think it makes sense for the requirement level to be associated with the attribute definition in the case of metrics. Instead it should be associated with the metric

I also agree with this. Another idea I had while reading your comments above: What if we split things in a way that you have: Metrics table and right below a table with its attribute? This way the requirement level can be defined there. 

The downside is the page becoming long and with duplication. But I think this improves readability. Maybe a TOC for each metric key would help navigate?

jamesmoessis commented:
@joaopgrassi yes, I like your idea, definitely makes it more readable. It means we need two tables for every metric, but I *think* that's a worthy sacrifice for the clarity it provides. I will have a shot at implementing this as soon as I get the chance!

Oberon00 commented:
@jamesmoessis @joaopgrassi I would prefer doing that improvement in a follow-up. If I understand correctly, it affects only the output, not the YAML structure?

joaopgrassi commented:
No problem in it being a follow up for me but I'll defer to @jamesmoessis to tell if there's no need to change the YAML structure.

jamesmoessis commented:
@Oberon00 @joaopgrassi sorry for not being active in the PR, I've been super snowed under with other work! 

The current yaml structure would need alterations in order to achieve the discussed structure. That being said, I'm not against merging this and then making those alterations in a follow up PR.

jamesmoessis commented:
The new YAML will look like this. There is one parent and each metric has it's own group as a direct child of the parent. It's simpler, easier to digest, more verbose, and we can generate the discussed structure far more easily. 
```
groups:
  - id: metric.http
    prefix: http
    type: metric
    brief: "This document defines semantic convention for HTTP client and server metrics."
    note: >
        These conventions can be used for http and https schemes
        and various HTTP versions like 1.1, 2 and SPDY.

  - id: metric.http.client.duration
      prefix: http
      type: metric
      extends: metric.http
      name: http.client.duration
      brief: "Measures the duration of the outbound HTTP request."
      instrument: Histogram
      units: ms
      attributes:
        - ref: http.method
          requirement_level: required
        - ref: http.status_code
          requirement_level:
            conditionally_required: "if and only if one was received/sent."
        - ref: http.flavor
          requirement_level: recommended
        - ref: net.peer.name
          requirement_level: required
        - ref: net.peer.port
          requirement_level:
            conditionally_required: If not default (`80` for `http`, `443` for `https`).
        - ref: net.sock.peer.addr
          requirement_level: recommended
```

I've committed a test file, I will now update the code so it can generate what we need.

jamesmoessis commented:
Ok, I've updated the code so it understands the new YAML structure. Take a look at `metrics.yaml` and what it turns into (`expected.md`) which follows our discussed markdown structure. I believe this is a great improvement on what we had before 😄

joaopgrassi commented on semantic-conventions/src/tests/data/markdown/metrics_tables/expected.md:
```diff
@@ -0,0 +1,15 @@
+| -------- | --------------- | ----------- | -------------- |
+| `foo.size` | Histogram | `{bars}` | Measures the size of foo. |
+<!-- endsemconv -->
+**Attributes for `foo.size`**
```
_Maybe_ an empty line between metric table and attributes table makes the .md file easier to read/review?

jamesmoessis replied:
Done :)

joaopgrassi commented on semantic-conventions/src/tests/data/markdown/metrics_tables/http.yaml:
```diff
@@ -0,0 +1,172 @@
+groups:
+  - id: http
```
Is this file here only needed because you use `ref` in `metrics.yaml` to refer to attributes from here? If so, maybe we can make it smaller by only adding the attributes we need?

jamesmoessis replied:
Thanks for pointing this out. I actually don't need this file anymore after the structure changed. I've removed it entirely and used the existing `http.yaml`.

joaopgrassi reviewed (commented):
Looks great! Thanks putting the effort and replying quickly. 

Two things we are missing: Update the [schema file](https://github.com/open-telemetry/build-tools/blob/main/semantic-conventions/semconv.schema.json#L130) to include the new properties for `MetricSemanticConvention`. and an entry in `CHANGELOG.md`

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +174,20 @@ The following is only valid if `type` is `event`:
+  - `id`, the ID of the metric. The fully qualified name of the metric includes its parent 
+    semantic convention prefix like so: `{parent.prefix}.{metric.id}`. 
+  - `brief`, a brief description of the metric.
+  - `instrument`, the [instrument type]( https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-types) 
```
It seems to me like the spec ought to have a section that is a good fit here. If it doesn't, that might be worth a spec issue/PR? The top-level TOC of your first link lists all instruments as sub-sections of the section you linked.

joaopgrassi replied:
Yeah I felt the same when I was reviewing.

> The top-level TOC of your first link lists all instruments as sub-sections of the section you linked.

You mean this? https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#meter ? It's still a lot of text and there's the "General characteristics" which just gets in the way. 

Maybe we really need a section that only explains the instruments. I can try opening an issue to at least ask about it.

jamesmoessis replied:
@joaopgrassi I think your first link to the API spec is the most appropriate. Good idea to list the instrument types, I will do that too.

joaopgrassi replied:
I just left a small comment but it looks good now. I can't resolve my own conversations for some reason, so feel free to do so.

jamesmoessis commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -247,7 +247,48 @@ def __init__(self, group):
+    yaml_to_markdown_instrument_repr: Dict[str, str] = {
+        "counter": "Counter",
+        "updowncounter": "UpDownCounter",
+        "histogram": "Histogram",
```
The exponential histogram isn't an instrument, rather it's a method of aggregation. Notice how the [API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument) doesn't mention exponential histograms. This is because the exponential histogram is purely an SDK/OTLP concept, where users would specify how to aggregate/export measurements given by instruments. 

Since auto-instrumentation only has access to to API, this means that the semantic convention can't and shouldn't specify aggregation types, this is left up to whoever configures the SDK.

joaopgrassi replied:
True, I missed that. All good, you can resolve this :)

jamesmoessis commented on semantic-conventions/syntax.md:
```diff
@@ -119,8 +119,14 @@ span_kind ::= "client"
+metric_name ::= string
 
+instrument ::=  "counter" 
+            | "histogram" 
```
See other comment

jamesmoessis commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +174,20 @@ The following is only valid if `type` is `event`:
+  - `instrument`, the [instrument type]( https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-types) 
+  that *should* be used to record the metric.
+  - `units`, the units in which the metric is measured, which should adhere to 
+     [UCUM](https://ucum.org/ucum.html). 
```
Good idea, I've changed the link.

jamesmoessis commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -247,7 +247,48 @@ def __init__(self, group):
+    def __init__(self, group):
+        super().__init__(group)
+        self.name = group.get("metric_name")
+        self.units = group.get("units")
```
Yeah, you're right. I've changed it.

jamesmoessis commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -247,7 +247,48 @@ def __init__(self, group):
+        if not all(val_tuple) and any(val_tuple):
+            raise ValidationError.from_yaml_pos(
+                self._position,
+                "Either all or none of name, units, and instrument must be defined",
```
The case where none of them are defined are the "parent" metrics in the yaml which aren't actually metrics. For example, in the test `metric.foo` is one of these. 

These exist for two reasons: 
- So multiple metrics can have a common parent and follow the same inheritance pattern that spans follow. 
- So attributes that aren't defined elsewhere can be defined on the parent without tying them to a specific metric.

jamesmoessis replied:
To clarify a bit more, this is the same as what happens in the span semantic convention currently. For example, the [network group](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/general.yaml#L2) exists so it can group attributes related to networks.

joaopgrassi replied:
Ah makes sense, got it. Thanks for the explanation, you can resolve this conversation :)

jamesmoessis commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -186,6 +189,61 @@ def to_markdown_attr(
+            )
+
+        output.write(
+            "| Name     | Instrument Type | Unit (UCUM) | Description    |\n"
```
For me, we should pick one. Having to define the unit twice in the YAML seems superfluous. I feel like `By` is pretty obvious, particularly when paired with the metric description. Also, any non-UCUM units like `{requests}` are also pretty obvious. 

Perhaps if we want to add this later on, it can be in another PR?

joaopgrassi replied:
Fair enough, let's go with this for now!

jamesmoessis commented:
Thanks for the review @joaopgrassi! I think I've addressed all your comments 😃

joaopgrassi commented on semantic-conventions/syntax.md:
```diff
@@ -178,15 +178,12 @@ The following is only valid if `type` is `event`:
+  - `metric_name`, the metric name as described by the [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#timeseries-model). 
+  - `instrument`, the [instrument type]( https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument) 
+  that *should* be used to record the metric. These include `counter`, `gauge`, `updowncounter`, and `histogram`. Whether the instrument 
+  is synchronous or asynchronous is an implementation detail and should not be specified. 
```
I feel the link before (https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-types) was great in clarifying why we only list the sync instrument types and now we are missing this. 

What about we write something like this instead? (to replace the text starting at `These include `counter`...`)

```
Note that the semantic conventions must be written using the synchronous instrument types
 (`counter`, `gauge`, `updowncounter` and `histogram`). For more details:
 [Metrics semantic conventions - Instrument types](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-types)
```

Or

```
Note that the semantic conventions must be written using the synchronous instrument types
 (`counter`, `gauge`, `updowncounter` and `histogram`). Whether implementations of these
conventions choose the synchronous or asynchronous instrument types 
is considered an implementation detail detail and should not be specified as part of these conventions.
```

jamesmoessis replied:
Yeah good point, I agree. I like the first example you gave since it links back to the source of truth. 

I've gone ahead and used that, with a slight modification/clarification: "Note that the semantic conventions must be written using **the names of** the synchronous instrument types".

I wanted to clarify that we are using the _names_ of the instrument types, it doesn't mean the implementation has to use the synchronous type. I think that makes it slightly clearly. Besides that, everything is the same as your example! Thanks for that 😄

joaopgrassi commented:
Hey @jmacd @reyang @osherv since you approved this PR, @jamesmoessis did some (big?) changes in the yaml structure and thus how it is rendered in markdown. Sorry for pinging but I just wanted to raise awareness and to make sure you still agree with it. Please ping others that you feel need to look at this as well. 

The gist of the changes is: Now we will have one metric table followed by its attributes **per metric**. You can check both the `metrics.yaml` and `expected.md` to see what it going to look like.

jamesmoessis commented:
@Oberon00 given the state of this PR, do you think it is ready for merge? We've left plenty of room for people to comment and voice any opinions, but haven't heard much. Personally I believe the PR is finally in a good state and ready to be merged.

Oberon00 commented:
The PR was completely changed at https://github.com/open-telemetry/build-tools/pull/79#issuecomment-1241517215. So I guess I should look at it again, as my review was only for an older version. The same goes for many comments.

joaopgrassi commented:
I will join the spec meeting this week and make sure folks are aware of this PR. Hopefully after that we should get more reviews.

joaopgrassi commented:
Hey @jamesmoessis can you please re-request a review from all the folks that approved before the "redesign" of the PR? I joined the spec meeting on 27.09 and that was a recommendation they gave us.

jamesmoessis commented:
Hi @tigrannajaryan - in the [27/9 spec SIG](https://youtu.be/cXYkWNo10vc?t=357) you mentioned you could remove the approvals that happened before the major changes I made. I've tried re-requesting review but when I do that for more than one person GitHub removes the prior request. Are you able to remove the approvals from those that approved before the major changes, and re-request their review? I believe those folks were @jmacd, @Oberon00, @reyang and @osherv

Oberon00 commented on semantic-conventions/README.md:
```diff
@@ -83,6 +83,9 @@ convention that have the tag `network`.
 `<!-- semconv http.server(tag=network, full) -->` will print the constraints and attributes of both `http` and `http.server`
 semantic conventions that have the tag `network`.
 
+`<!-- semconv metric.http.server(metric_table) -->` will print a table of metrics with all metrics prefixed with 
```
Is this really depending on the prefix? Probably on the group ID?

jamesmoessis replied:
Good catch, this example should've changed when I made the changes to the PR. 

```
`<!-- semconv metric.http.server.active_requests(metric_table) -->` will print a table describing a single metric 
`http.server.active_requests`. 
```

The group ID is `metric.http.server.active_requests` but the `metric_name` (a new field I added) will be printed to the markdown, which in this case would be `http.server.active_requests`.

arminru commented on semantic-conventions/src/tests/data/markdown/metrics_tables/expected.md:
```diff
@@ -0,0 +1,16 @@
+# Test Markdown
+
+**`foo.size`**
+<!-- semconv metric.foo.size(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `foo.size` | Histogram | `{bars}` | Measures the size of foo. |
+<!-- endsemconv -->
```
Can you please add a second metric to demonstrate how the table would be rendered here?
It would be one single table defining name+instrument+unit+description for all metrics and then below one table _each_ for the respective attributes, right?

jamesmoessis replied:
> Can you please add a second metric to demonstrate how the table would be rendered here? 

Sure, I agree this would make it clearer. I will add it in. 

> It would be one single table defining name+instrument+unit+description for all metrics and then below one table each for the respective attributes, right?

Take a look at the discussion around this comment: https://github.com/open-telemetry/build-tools/pull/79#issuecomment-1193448303

We decided that: 
* A "metric table" only has one metric. This means that each metric has it's own single-row table describing it's name, unit, description, and instrument. 
* Below a metric table will be an attribute table, associated with the above metric.

jamesmoessis replied:
I've added another metric and attribute 👍

arminru replied:
I think we could also define all metrics in one metrics table at once, and then have the respective attributes in their own table like this. This would still allow individual requirement levels:

```
## foo metrics
<!-- semconv metric.foo(metric_table) -->
| Name     | Instrument Type | Unit (UCUM) | Description    |
| -------- | --------------- | ----------- | -------------- |
| `foo.size` | Histogram | `{bars}` | Measures the size of foo. |
| `foo.duration` | Histogram | `{bazs}` | Measures the duration of foo. |
<!-- endsemconv -->

### Attributes for `metric.foo.size`
<!-- semconv metric.foo.size -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `http.method` | string | HTTP request method. | `GET`; `POST`; `HEAD` | Required |
| `http.status_code` | int | [HTTP response status code](https://tools.ietf.org/html/rfc7231#section-6). | `200` | Conditionally Required: if and only if one was received/sent. |
<!-- endsemconv -->

### Attributes for `metric.foo.duration`
<!-- semconv metric.foo.duration -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `http.method` | string | HTTP request method. | `GET`; `POST`; `HEAD` | Optional |
<!-- endsemconv -->
```

resulting in

---

## foo metrics

| Name     | Instrument Type | Unit (UCUM) | Description    |
| -------- | --------------- | ----------- | -------------- |
| `foo.size` | Histogram | `{bars}` | Measures the size of foo. |
| `foo.duration` | Histogram | `{bazs}` | Measures the duration of foo. |

### Attributes for `metric.foo.size`

| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `http.method` | string | HTTP request method. | `GET`; `POST`; `HEAD` | Required |
| `http.status_code` | int | [HTTP response status code](https://tools.ietf.org/html/rfc7231#section-6). | `200` | Conditionally Required: if and only if one was received/sent. |

### Attributes for `metric.foo.duration`

| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `http.method` | string | HTTP request method. | `GET`; `POST`; `HEAD` | Optional |

---

This can still be discussed and (if agreed on doing so) improved later on. Let's get this PR here in first 🙂 
Unless I miss something this would not break the YAML definitions but would only affect the generated markdown, which we can easily adapt in future iterations.

jamesmoessis replied:
I like this idea, and will help make it more concise. I think we should definitely do this, but also agree this should be done in a later PR. You are right that we can do this with the current yaml structure.

Oberon00 replied:
I'm strongly against single-row tables. Instead of making it a table with multiple columns and a single row, why don't you make it a bullet-point list like

> **What would be the column header**: What would be the cell content below that header.

Alternatively, I think it would make more sense to have the `metric_table` be an argument to the metric_group's rendering, and then it would list all the metrics in that group. This is probably to be considered together with my comment about the `extends`-mechanism not really being ideal for declaring the association between metric and metric_group.

joaopgrassi replied:
If single-row tables is not desired, I think @arminru suggestion's on having the metrics together in a single table and splitting the attributes makes more sense. Not sure if I follow the bullet-point approach though, some examples would be nice to compare both approaches :)

jamesmoessis replied:
We all agree that single row tables aren't the way to go. We just don't want to extend the scope of this PR more. In a later PR, we can extend so it works for arbitrary amount of rows.

joaopgrassi replied:
Maybe we should just them agree on what will be the follow-up solution? Either the ones suggested by @arminru  or by @Oberon00, which I'm not sure I follow exactly.

Oberon00 replied:
My second suggested solution was that you reference the metric group in the markdown, and it then renders a single table with multiple rows, one for each metric instrument that uses that metric group.

jamesmoessis replied:
If I understand correctly, I think both @arminru and @Oberon00 suggestions are the same or very similar. The `metric.foo` group in @arminru's example is the metric group that @Oberon00 is talking about where multiple metrics are listed. 

I definitely agree we should group metrics into a table and have separate attribute tables for each metric. I also think that we agree that this can be done in a later PR. This PR is now very bloated and reaches it's 1 year birthday in 2 days time.

jsuereth replied:
I agree that we should expand cleanup here later.  At this point, we just need something that works so we can define metric semantic conventions and share definitions with traces.  The most important aspect of this PR is *not* how it generates tables, which we can fix over time, but how the YAML is defined and shared between signals.  Let's focus on that as the most important part of this PR.




Also, regarding table generation, there are currently two main styles that exist *in our specification*:

1.  A table of metric names with a (then shared) table of attribute names that apply.
2.  A table of the metric where "empty" rows are created while attribtues are expanded (I didn't see this proposed).

I think #2 joined with what @jamesmoessis has w/ one-metric table kind of makes sense as you can see the cardinality of a metric pretty easily then.

However, let's continue tweaking how metric tables can be generated in a *follow up PR*

arminru commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +174,18 @@ The following is only valid if `type` is `event`:
+  For more details: [Metrics semantic conventions - Instrument types](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-types).
+  - `unit`, the unit in which the metric is measured, which should adhere to 
+    [the guidelines](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/metrics/semantic_conventions#instrument-units). 
+
```
You could add that all three properties are required and maybe that optional attributes can be added.

jamesmoessis replied:
They aren't technically required, due to what I said over here: https://github.com/open-telemetry/build-tools/pull/79#discussion_r999985536 

They are required for "concrete" metrics, but for the parent metrics, they aren't.

arminru replied:
see https://github.com/open-telemetry/build-tools/pull/79#discussion_r1003468865

arminru replied:
Closing - this was resolved in the meantime.

arminru commented on semantic-conventions/src/tests/data/yaml/metrics.yaml:
```diff
@@ -0,0 +1,22 @@
+    note: >
+        Details about foo.
+
+  - id: metric.foo.size
+    prefix: foo
+    type: metric
+    extends: metric.foo
+    metric_name: foo.size
```
What is the `prefix` used for if `metric_name` is already the fully-qualified name? I'd either expect the prefix to not be used for metrics at all or that it would be prepended to the `metric_name` (which would then only be "size").

jamesmoessis replied:
`prefix` is used for the attribute names, not the metric names. 

I admit it is confusing from the current test. I will add in an attribute definition to the test so it's more clear how it works.

jamesmoessis replied:
Agreed it could be clearer 👍 I've gone ahead and made the `prefix: bar`.  See https://github.com/open-telemetry/build-tools/pull/79/commits/6a67c74f478f2dc70c07ae2c62183e78110c1e1e.

arminru commented on semantic-conventions/src/tests/data/yaml/metrics.yaml:
```diff
@@ -0,0 +1,22 @@
+groups:
+  - id: metric.foo
+    prefix: foo
+    type: metric
+    brief: "This document defines foo."
+    note: >
+        Details about foo.
+
```
This is non-functional, right? It looks as if it would define a metric but there is no instrument or unit for it.

jamesmoessis replied:
It serves a purpose of acting as a parent for the other, "concrete" metrics, essentially the root of a tree where related metrics can be grouped under. It's somewhere that attributes which are associated with multiple child metrics can be defined under. 

This is similar to how it works in the trace semconv, for example [http.yaml](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/http.yaml#L2) has a top-level "http" group which serves as a parent for other spans, while being a place that attributes for all of http can be defined.

arminru replied:
In that case, the `http` entity of `type: span` does describe parts of an actual span (its attributes), however. I find the dual use of entities of `type: metric` a bit surprising.

Do you think adding a dedicated `type` called `group` or `metric_group` would make sense?
Then all entities of type `metrics` would be an actual metric, consisting of a `metric_name`, `instrument` and `unit`, while still being put under a common root.
Or we leave out the root and have only the individual metrics defined, while they are still combined in the same YAML file.

jamesmoessis replied:
> Do you think adding a dedicated type called group or metric_group would make sense?

Yes, this makes sense, and it was a very simple change to make. See https://github.com/open-telemetry/build-tools/pull/79/commits/e524ab454f344692efe22d90d9beb78338a1edd6. I believe the distinction should be clear now.

arminru reviewed (approved):
With the latest clarifications and improvements, particularly e524ab454f344692efe22d90d9beb78338a1edd6, this looks good to me!

The resulting rendered Markdown might be improved as suggested in https://github.com/open-telemetry/build-tools/pull/79#discussion_r1000834823 but this can be iterated on later.

Thank you for your contribution, @jamesmoessis!

joaopgrassi commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -244,7 +244,11 @@ def __init__(self, group):
 
-class MetricSemanticConvention(BaseSemanticConvention):
+class MetricGroupSemanticConvention(BaseSemanticConvention):
+    GROUP_TYPE_NAME = "metric_group"
```
I think also the schema.json needs updating. It still references the old class name.

arminru replied:
Ah yes, good catch! No, it's not just internal. If configured correctly, vscode for instance will use it to validate your YAMLs when editing the OTel spec repo.

Oberon00 replied:
And give you autocompletions, which is quite nice.

jamesmoessis replied:
Thankyou for reminding! I have updated it with my recent commit https://github.com/open-telemetry/build-tools/pull/79/commits/4a8aabdb6a0a3429a41566043a0bfb9783229064

jamesmoessis commented:
@Oberon00 is there anything else left here or can we go ahead and merge?

Oberon00 commented:
I'll try to review this today, but if I don't get to it it's fine with me to merge it without my review.

Oberon00 commented on semantic-conventions/CHANGELOG.md:
```diff
@@ -4,7 +4,8 @@ Please update the changelog as part of any significant pull request.
 ## Unreleased
 
-- ...
+- Add definitions for metrics in YAML semconv, plus generation to markdown
```
You don't add defintions for metrics, you add support for them. Maybe copy the wording from the changelog entry below which added the "scope" type.

jamesmoessis replied:
Done

Oberon00 commented on semantic-conventions/README.md:
```diff
@@ -83,6 +83,9 @@ convention that have the tag `network`.
 `<!-- semconv http.server(tag=network, full) -->` will print the constraints and attributes of both `http` and `http.server`
 semantic conventions that have the tag `network`.
 
+`<!-- semconv metric.http.server.active_requests(metric_table) -->` will print a table describing a single metric 
```
What happens when you leave out `metric_table`? Can this be auto-detected?

Oberon00 commented on semantic-conventions/src/tests/data/yaml/metrics.yaml:
```diff
@@ -0,0 +1,42 @@
+  - id: metric.foo.size
+    prefix: foo
+    type: metric
+    extends: metric.foo
```
Does the "extends" mechanism really make sense here? I guess it's fine as a prototype, but it will always copy all the attributes, and the brief/etc.

jamesmoessis replied:
Yeah, fair point. It wasn't necessary here. All tests still pass without it. I've removed the extends here, it was superfluous.

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +178,26 @@ The following is only valid if `type` is `event`:
+#### Metric Group semantic convention
+
+Metric group inherits all from the base semantic convention, and does not 
+add any additional fields. Its only requirement is that `type` is `metric_group`. 
```
```suggestion
add any additional fields.
```

I think this is so obvious that stating it explicitly is confusing.

jamesmoessis replied:
Done

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +178,26 @@ The following is only valid if `type` is `event`:
+add any additional fields. Its only requirement is that `type` is `metric_group`. 
+
+The metric group semconv is intended to be a group where metric attributes 
+can be defined and then referenced from other `metric` groups. 
```
Referenced how?

jamesmoessis replied:
Using `ref`, the same as span attributes do. I've clarified.

Oberon00 commented on semantic-conventions/syntax.md:
```diff
@@ -168,6 +178,26 @@ The following is only valid if `type` is `event`:
   If not specified, the `prefix` is used. If `prefix` is empty (or unspecified),
   `name` is required.
 
+#### Metric Group semantic convention
```
I'm missing a statement of the form "A metric group is ...", i.e. what is a metric group conceptually.

jamesmoessis replied:
After some rewording, this section now includes: 

> The metric group semconv is a group where related metric attributes can be defined and then referenced from other `metric` groups using `ref`.

Does it make sense?

Oberon00 replied:
Would this still be needed after #124, or would these be redundant?

arminru replied:
@jamesmoessis @lmolkova #124 should be unrelated, right?
One is for grouping metrics (to have an identifier to print out related ones in one table) and the other is for attribute groups (so they can be shared as a whole between different signals).

Oberon00 replied:
#124 can be used for sharing as a whole or part by part. It can also be used to group related attributes and a table can be generated from it on its own IIUC.

arminru replied:
For grouping entire metric definitions - consisting of generic fields (including attributes) and the metric-specific fields metric_name, instrument and unit - the metric_group would be used.

So I think for the combined table generation that will be done in a follow-up PR as per the thread in https://github.com/open-telemetry/build-tools/pull/79/files#r996803040 this would lay the foundation.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -244,10 +244,55 @@ def __init__(self, group):
+        "instrument",
+    )
+
+    yaml_to_markdown_instrument_repr: Dict[str, str] = {
```
Ideally, this should be placed in the markdown generator, if it's really about the markdown representation only. Or name it differently, like `canonical_instrument_name_by_yaml_name`

jamesmoessis replied:
I've changed the name to the one you suggested. I've also avoided storing the `self.instrument_markdown_fmt` and instead just calculate it using this dict during the mardkown rendering. 

I think keeping this dict makes sense, because it also defines the allowed yaml values, while mapping them to their markdown values. Otherwise I would have to redefine this whole dict again in the markdown renderer.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/model/semantic_convention.py:
```diff
@@ -244,10 +244,55 @@ def __init__(self, group):
+        self.metric_name = group.get("metric_name")
+        self.unit = group.get("unit")
+        self.instrument = group.get("instrument")
+        self.instrument_markdown_fmt = self.yaml_to_markdown_instrument_repr.get(
```
Maybe do this after `validate()` and use `[]` instead of `.get`

jamesmoessis replied:
I've removed this line, see https://github.com/open-telemetry/build-tools/pull/79#discussion_r1035414725

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -186,6 +189,61 @@ def to_markdown_attr(
             )
         )
 
+    def to_markdown_attribute_table(
```
Were there any changes to code here, besides extracting it?

jamesmoessis replied:
No changes to this function, looking at the git blame I haven't touched it since April

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -186,6 +189,61 @@ def to_markdown_attr(
+        attr_sampling_relevant = [
+            attr for attr in attr_to_print if attr.sampling_relevant
+        ]
+        self.to_creation_time_attributes(attr_sampling_relevant, output)
```
This will change the output by reordering the paragraphs I think.

jamesmoessis replied:
Yeah, I think it doesn't matter too much.

jsuereth commented:
@jamesmoessis  What's the status on this?  Do you need any assistance fixing the changes suggested by @Oberon00 ?  I'd love if we could get this merged soon.  Let me know if a PR against your PR with fixes would be appreciated or if you'll have time to address this soon.

jamesmoessis commented:
Ok @jsuereth @joaopgrassi @Oberon00 I think I've addressed all comments. 

The build is failing pending on this fix regarding the linter: https://github.com/open-telemetry/build-tools/pull/125

arminru commented:
@jamesmoessis Thank you! I merged #125 but I think you'll have to update your branch manually to resolve the conflicts.

jamesmoessis commented:
Thanks @arminru - I've resolved the conflicts.

Oberon00 commented on semantic-conventions/src/opentelemetry/semconv/templating/markdown/__init__.py:
```diff
@@ -216,16 +214,15 @@ def to_markdown_metric_table(
             "| -------- | --------------- | ----------- | -------------- |\n"
         )
         output.write(
+            # pylint: disable=C0209
```
Instead of disabling, maybe store the complex expression in a named variable to make the code a bit more clear. If disabling, please don't disable by code, but with the readable identifier.

jamesmoessis replied:
Good idea, Done.

jamesmoessis commented:
@arminru @jsuereth @Oberon00 @joaopgrassi Any other comments or can we go ahead and merge?

joaopgrassi commented:
@jamesmoessis not sure if the open comments are already resolved, but on my part, I think we need to agree on what approach we will take on the table rendering:

https://github.com/open-telemetry/build-tools/pull/79#discussion_r1035791763

So we have a plan for the follow up.

jsuereth commented:
@joaopgrassi I actually don't think we need perfect table rendering yet.

The most important aspect of this PR is that we can define metric semantic conventions in YAML and generate *any* form of documentation for them.  I think what @jamesmoessis has now is usable, and tweakable going forward.  If we want to make improvements, I suggest folks that want different views tackle that work.

For now, we really need *any* functionality that lets us document metric semantic conventions programatically and share attributes w/ traces & logs.  If we have concerns for how that works, I'd consider it blocking.

Let's not bikeshed too much on Markdown tables, especially when those are easy to fix once we have data in YAML.   Right now we don't even have our semconv in YAML.

marcalff commented:
> @joaopgrassi I actually don't think we need perfect table rendering yet.
> 

+1 on that.

From the language implementation point of view (like opentelemetry-cpp), what we _need_ is:

- metrics semantic conventions represented in yaml files,
- build tools to generate code, for metrics.

Please don't block progress because of rendering issues on table.md, if it can be resolved independently (and therefore, in parallel).

joaopgrassi commented:
Yes, I agree! All my comments were "not blocking" (sorry if it was unclear) and I have been advocating to merge this so we have _something._ 

I just wanted to make sure there's a consensus going forward so two people don't do double work on the follow up.

arminru commented:
I went through the remaining unresolved conversations and reached out to the authors to resolve them.

I'll also resolve the last one - https://github.com/open-telemetry/build-tools/pull/79/files#r996803040 - as we agreed that this lays a solid foundation for defining metrics in YAML and that the MD rendering can be tackled in a follow-up PR. I created #129 for that.

Thank you a lot for your work and endurance, @jamesmoessis! 😊 
Also thanks to the reviewers for the multiple rounds of input, this looks good now!
